reconcile_test.go

```diff
@@ -113,223 +66,234 @@ func TestReconcile(t *testing.T) {
+				},
+				Voters: []raft.ServerID{
+					"11111111-1111-1111-1111-111111111111",
+					"22222222-2222-2222-2222-222222222222",
+					"33333333-3333-3333-3333-333333333333",
+					"44444444-4444-4444-4444-444444444444",
+				},
+			},
+			changes: RaftChanges{
+				Promotions: []raft.ServerID{
+					// both promotions should happen
+					"55555555-5555-5555-5555-555555555555",
+					"66666666-6666-6666-6666-666666666666",
+				},
+				Demotions: []raft.ServerID{
+					// Both servers are unhealthy, but no demotions should happen in this
+					// reconciliation since we are starting with an even number of voters. We're
+					// still passing the demotions to the reconcile function to ensure that they are
+					// filtered out.
+					"33333333-3333-3333-3333-333333333333",
+					"44444444-4444-4444-4444-444444444444",
+				},
+			},
+			setupExpectations: func(m *MockRaft) {
+				m.On("AddVoter",
+					raft.ServerID("55555555-5555-5555-5555-555555555555"),
+					raft.ServerAddress("198.18.0.5:8300"),
+					uint64(0),
+					time.Duration(0),
+				).Return(&raftIndexFuture{}).Once()
+				m.On("AddVoter",
+					raft.ServerID("66666666-6666-6666-6666-666666666666"),
+					raft.ServerAddress("198.18.0.6:8300"),
+					uint64(0),
+					time.Duration(0),
+				).Return(&raftIndexFuture{}).Once()
+			},
+		},
+		"demote-any-with-even-voters-and-no-promotions": {
+			state: State{
+				Leader: "11111111-1111-1111-1111-111111111111",
+				Servers: map[raft.ServerID]*ServerState{
+					"11111111-1111-1111-1111-111111111111": makeServerState("1", RaftLeader, true),
+					"22222222-2222-2222-2222-222222222222": makeServerState("2", RaftVoter, true),
+					"33333333-3333-3333-3333-333333333333": makeServerState("3", RaftVoter, false),
+					"44444444-4444-4444-4444-444444444444": makeServerState("4", RaftVoter, false),
+					"55555555-5555-5555-5555-555555555555": makeServerState("5", RaftVoter, true),
+					"66666666-6666-6666-6666-666666666666": makeServerState("6", RaftVoter, true),
+				},
+				Voters: []raft.ServerID{
+					"11111111-1111-1111-1111-111111111111",
+					"22222222-2222-2222-2222-222222222222",
+					"33333333-3333-3333-3333-333333333333",
+					"44444444-4444-4444-4444-444444444444",
+				},
+			},
+			changes: RaftChanges{
+				Demotions: []raft.ServerID{
+					"33333333-3333-3333-3333-333333333333",
+					"44444444-4444-4444-4444-444444444444",
+					"55555555-5555-5555-5555-555555555555",
+				},
+			},
+			setupExpectations: func(m *MockRaft) {
+				m.On("DemoteVoter",
+					raft.ServerID("33333333-3333-3333-3333-333333333333"),
+					uint64(0),
+					time.Duration(0),
+				).Return(&raftIndexFuture{}).Once()
+				m.On("DemoteVoter",
+					raft.ServerID("44444444-4444-4444-4444-444444444444"),
+					uint64(0),
+					time.Duration(0),
+				).Return(&raftIndexFuture{}).Once()
+				m.On("DemoteVoter",
+					raft.ServerID("55555555-5555-5555-5555-555555555555"),
+					uint64(0),
+					time.Duration(0),
+				).Return(&raftIndexFuture{}).Once()
```

Comment:
It looked off to me at first to see us demoting a full half of the cluster in one round 😱 . But I guess this test is not setting a min_quorum so this is OK because it's actually doing each as a separate config change and reducing quorum as it goes, ending up leaving a cluster with 3 healthy servers in a quorum right?

Reply:
Yep, that was my thinking. These would all be done in a single reconcile loop (this is the behaviour from before this PR anyways), but separate raft reconfigurations. Mainly though, I wanted to ensure that we have a test case asserting on multiple demotions, regardless of health status, happening in a single reconcile round that starts with an even number of voters.